driver/labjack/writer.h

```diff
@@ -0,0 +1,222 @@
+    double state = 0.0;
+    synnax::DataType data_type = synnax::FLOAT64;
+    synnax::ChannelKey state_key = 0;
+};
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                    StateSource                                //
+///////////////////////////////////////////////////////////////////////////////////
+class StateSource final : public pipeline::Source {
+public:
+    explicit StateSource() = default;
+
+    explicit StateSource(
+        const synnax::Rate state_rate,
+        const synnax::ChannelKey &state_index_key,
+        const std::map<synnax::ChannelKey, labjack::out_state> state_map
+    );
+
+    std::pair<synnax::Frame, freighter::Error> read(breaker::Breaker &breaker) override;
+
+    synnax::Frame get_state();
+
+    void update_state(synnax::Frame frame);
+
+private:
+    std::mutex state_mutex;
+    std::condition_variable waiting_reader;
+    synnax::Rate state_rate = synnax::Rate(1);
+    std::map<synnax::ChannelKey, labjack::out_state> state_map;
+    synnax::ChannelKey state_index_key;
+    loop::Timer timer;
+}; // class StateSource
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                   WriterChannelConfig                         //
+///////////////////////////////////////////////////////////////////////////////////
+struct WriterChannelConfig {
+    std::string location;
+    bool enabled = true;
+    synnax::DataType data_type;
+    uint32_t cmd_key;
+    uint32_t state_key;
+    std::string channel_type = "";
+
+    WriterChannelConfig() = default;
+
+    explicit WriterChannelConfig(config::Parser &parser)
+        : enabled(parser.optional<bool>("enabled", true)),
+          data_type(parser.optional<std::string>("data_type", "uint8")),
+          cmd_key(parser.required<uint32_t>("cmd_key")),
+          state_key(parser.required<uint32_t>("state_key")),
+          channel_type(parser.optional<std::string>("channel_type", "")),
+          location(parser.optional<std::string>("port", "")) {
+        if (!parser.ok())
+            LOG(ERROR) << "Failed to parse writer channel config: " << parser.
+                    error_json().dump(4);
+    }
+};
+
+///////////////////////////////////////////////////////////////////////////////////
+//                                   WriterConfig                                //
+///////////////////////////////////////////////////////////////////////////////////
+struct WriterConfig {
+    std::string device_type;
+    std::string device_key;
+    std::vector<WriterChannelConfig> channels;
+    synnax::Rate state_rate = synnax::Rate(1);
+    std::string serial_number; // used to open devices
+    std::string connection_type;
+    bool data_saving;
+    std::string task_name;
+    synnax::ChannelKey task_key;
+    std::map<synnax::ChannelKey, labjack::out_state> initial_state_map;
+    synnax::ChannelKey state_index_key;
```

Comment:
There is no guarantee that there is only one state index key. There can be many state index keys

Reply:
ok. this is also currently how it works in the ni writer so ill have to change that too. im going to just first change it here tho